File: src/solution_java/medium/_16ThreeSumClosest.java
package solution_java.medium;

import java.util.Arrays;

public class _16ThreeSumClosest {
    public class Solution {
        public int threeSumClosest(int[] nums, int target) {
            if (nums.length == 3) {
                return nums[0] + nums[1] + nums[2];
            }
            int closeTarget = 100000;
            Arrays.sort(nums);
            int secondIndex;
            int thirdIndex;
            for (int i = 0; i < nums.length; i++) {
                if (i > 0 && nums[i] == nums[i - 1]) {
                    continue;
                }
                secondIndex = i + 1;
                thirdIndex = nums.length - 1;
                while (secondIndex < thirdIndex) {
                    int sum = nums[i] + nums[secondIndex] + nums[thirdIndex];
                    if (sum == target) {
                        return sum;
                    }
                    int delta = sum - target;
                    if (Math.abs(delta) < Math.abs(closeTarget - target)) {
                        closeTarget = sum;
                    }

                    if (sum > target) {
                        int tempThird = thirdIndex - 1;
                        while (secondIndex < tempThird && nums[tempThird] == nums[thirdIndex]) {
                            tempThird -= 1;
                        }
                        thirdIndex = tempThird;
                    } else {
                        int tempSecond = secondIndex + 1;
                        while (tempSecond < thirdIndex && nums[tempSecond] == nums[secondIndex]) {
                            tempSecond += 1;
                        }
                        secondIndex = tempSecond;
                    }
                }
            }
            return closeTarget;
        }
    }
}

File: src/medium/_16ThreeSumClosest.kt
package medium

import java.util.*

class _16ThreeSumClosest {
class Solution {
    fun threeSumClosest(nums: IntArray, target: Int): Int {
        if (nums.size < 3) {
            return 1000000
        }
        if (nums.size == 3) {
            return nums[0] + nums[1] + nums[2]
        }
        Arrays.sort(nums)
        var closestTarget = 1000000
        var secondPointIndex: Int
        var thirdPointIndex: Int
        for (x in nums.indices) {
            if (x > 0 && nums[x] == nums[x - 1]) {
                continue
            }
            secondPointIndex = x + 1
            thirdPointIndex = nums.size - 1
            while (secondPointIndex < thirdPointIndex) {
                val sum = nums[x] + nums[secondPointIndex] + nums[thirdPointIndex]
                if (sum == target) {
                    return sum
                }
                val delta = nums[x] + nums[secondPointIndex] + nums[thirdPointIndex] - target
                if (Math.abs(delta) < Math.abs(closestTarget - target)) {
                    closestTarget = sum
                }
                if (sum > target) {
                    var tempThird = thirdPointIndex - 1
                    while (tempThird > secondPointIndex && nums[tempThird] == nums[thirdPointIndex]) {
                        tempThird--
                    }
                    thirdPointIndex = tempThird
                } else {
                    var tempSecond = secondPointIndex + 1
                    while (tempSecond < thirdPointIndex && nums[tempSecond] == nums[secondPointIndex]) {
                        tempSecond++
                    }
                    secondPointIndex = tempSecond
                }
            }
        }
        return closestTarget
    }
}
}